src/items.rs

```diff
@@ -607,6 +607,112 @@ pub fn format_struct(context: &RewriteContext,
     }
 }
 
+pub fn format_trait(context: &RewriteContext, item: &ast::Item, offset: Indent) -> Option<String> {
+    if let ast::ItemKind::Trait(unsafety, ref generics, ref type_param_bounds, ref trait_items) =
+           item.node {
+        let mut result = String::new();
+        let header = format!("{}{}trait {}",
+                             format_visibility(item.vis),
+                             format_unsafety(unsafety),
+                             item.ident);
+
+        result.push_str(&header);
+
+        // TODO: Add max_width checking
+        // let budget = try_opt!(context.config.max_width.checked_sub(result.len()));
+        // let indent = offset + result.len();
+
+        let body_lo = context.codemap.span_after(item.span, "{");
+
+        let generics_str = try_opt!(rewrite_generics(context,
+                                                     generics,
+                                                     offset,
+                                                     context.config.max_width,
+                                                     offset + result.len(),
+                                                     mk_sp(item.span.lo, body_lo)));
+        result.push_str(&generics_str);
+
+        let trait_bound_str = try_opt!(rewrite_trait_bounds(context,
+                                                            type_param_bounds,
+                                                            offset,
+                                                            context.config.max_width));
+
+        if offset.width() + result.len() + trait_bound_str.len() > context.config.max_width {
+            result.push('\n');
+            let width = context.block_indent.width() + context.config.tab_spaces;
+            let trait_indent = Indent::new(0, width);
+            result.push_str(&trait_indent.to_string(context.config));
+        }
+        result.push_str(&trait_bound_str);
+
+        let where_budget = try_opt!(context.config.max_width.checked_sub(last_line_width(&result)));
+        let where_clause_str = try_opt!(rewrite_where_clause(context,
+                                                             &generics.where_clause,
+                                                             context.config,
+                                                             context.config.item_brace_style,
+                                                             context.block_indent,
+                                                             where_budget,
+                                                             context.config.where_density,
+                                                             "{",
+                                                             context.config.where_trailing_comma,
+                                                             None));
+        if !where_clause_str.contains('\n') &&
```

Comment:
where clause should always be on a newline unless we can fit the whole trait on one line. For that you'll also need to test that the body is empty and that result does not contain `\n`, as well as the tests you have so far.

You can leave this for later and always insert a newline before the where clause if you like.